File: java/Choice.java
package irresponsible.crouton;

import clojure.lang.IPersistentVector;
import clojure.lang.IPersistentMap;
import java.util.List;
import java.util.ArrayList;

public class Choice implements IRoute {
  private final ArrayList<IRoute> routes;
  public Choice(List<IRoute> choices) {
    if (null == choices)
      throw new IllegalArgumentException("choices must not be nil");
    int count = choices.size();
    if (count == 0)
      throw new IllegalArgumentException("choices must contain at least one item");
    routes = new ArrayList<IRoute>(choices);
  }
  public final Object match(IPersistentVector pieces, IPersistentMap places) {
    Object ret = null;
    for(IRoute route : routes) {
      ret = route.match(pieces, places);
      if (ret != null) return ret;
    }
    return null;
  }
}

File: kotlin/src/Choice.kt
package irresponsible.crouton

import clojure.lang.IPersistentVector
import clojure.lang.IPersistentMap
import java.util.ArrayList

class Choice(choices: List<IRoute>) : IRoute {
    private val routes = ArrayList(choices)

    init {
      if (choices.size == 0) throw IllegalArgumentException("choices must contain at least one item")
    }

    override fun match(pieces: IPersistentVector, places: IPersistentMap): Any? {
        for (route in routes) {
          val ret = route.match(pieces, places)
	  if (ret != null) return ret
        }
        return null
    }
}
